### PR TITLE
Check for trailing garbage inside json (rework #80)

### PR DESCRIPTION
**Pull request by DimitriPapadopoulos**
Initial attempt to rework and understand #80.

### CODE REVIEW

**Comment by DimitriPapadopoulos** [comment]
I have squashed the first commits from #80:
* 0b89a1c `Check for trailing garbage inside json`
* b3e4bc9 `cleanup`
* 06ac53f `more cleanup`
* 32524f5 `indentation`
* 86d646c `formatting`

and left out the last commit (for now) because it is the one that introduces unwanted formatting changes:
* d63b9d8 `Allow for empty objects and empty arrays`

**Comment by LB--** [comment]
Thanks for cleaning it up. It's still unclear to me what this checks for or why it is so complicated. We should first create the test JSON files that fail in order to understand what a solution would be.

Also, I'm sort of concerned that the exclusion of the "Allow for empty objects and empty arrays" commit hasn't resulted in a test failure... do we not have tests with empty objects and empty arrays?

**Comment by DimitriPapadopoulos** [comment]
Still unclear to me too! I'm cleaning up the patch in the hope of understanding what #80 is supposed to fix. Then we can write a relevant test. Then perhaps we can suggest simpler code.

File: json.c [hunk 1]
```diff
@@ -1053,3 +1070,40 @@ void json_value_free (json_value * value)
...
+
+   do {
+      ptr--;
+   } while (isspace(*ptr));
+
+   switch (*ptr)
+   {
+      case '}':
+      case ']':
+      case '"':
```
Comment (tmegow) [inline json.c]:
```suggestion
       case '}':
       case '{':
       case ']':
       case '[':
       case '"':
```

**Review by tmegow — COMMENTED** [review]
@DimitriPapadopoulos I see your comment [here](https://github.com/json-parser/json-parser/issues/73#issuecomment-962441740) about parsers being more lenient than linters. I didn't realize this is the case; today I learned this from you! 

My goal with these changes is to parse AND lint `ECMA-404` JSON which doesn't strictly accept trailing commas (like the example below).

I appreciate your creating this cleaned-up PR. Also, thank you for the review + suggestions on #80. I will happily add your suggestions. Can I be given permission to edit this branch?


> Still unclear to me too! I'm cleaning up the patch in the hope of understanding what #80 is supposed to fix. Then we can write a relevant test. Then perhaps we can suggest simpler code.

Does these logs help clarify my goal?

```
$ cat ../tests/invalid-0011.json # an example relevant test
{
    "a": "b",
    "c": "d",
    "foo": [
        "a",
    ],
    "trailing-comma": {},
}

$ git checkout master && gcc -o test_json -I.. test_json.c ../json.c -lm && ./test_json ../tests/invalid-0011.json # This parses successfully with trailing commas in objects/arrays.

Already on 'master'
Your branch is up to date with 'origin/master'.
{
    "a": "b",
    "c": "d",
    "foo": [
        "a",
    ],
    "trailing-comma": {},
}

--------------------------------

 object[0].name = a
  string: b
 object[1].name = c
  string: d
 object[2].name = foo
  array
   string: a
 object[3].name = trailing-comma

$ git checkout rework && gcc -o test_json -I.. test_json.c ../json.c -lm && ./test_json ../tests/invalid-0011.json # I desire this to fail because linting fails

Switched to branch 'rework'
{
    "a": "b",
    "c": "d",
    "foo": [
        "a",
    ],
    "trailing-comma": {},
}

--------------------------------

Unable to parse data
```

To this review, I added the content from the `d63b9d8 Allow for empty objects and empty arrays` commit which was excluded from the Draft PR. These lines fix the fact that the `trailing_garbage` function+invocation broke empty object/array parsing. Also `trailing_garbage` can probably be renamed to `trailing_commas` (`%s/trailing_garbage/trailing_commas/g`) because the only reserved character which can wind up at the end of objects/arrays and end up with this successful parsing.

Because I desire a version of this parser which also does this particular linting, have you already thought of a different (improved) strategy to achieve this goal? I am curious to see any improvements on my attempt.

**Comment by DimitriPapadopoulos** [comment]
I think it's best if you can :
* remove all formatting changes from the `master` branch of [`tmegow/json-parser`](https://github.com/tmegow/json-parser), so that the intent of the patch is clear and not lost in the noise from the formatting changes,
* then move these changes to a specific branch in your repo, such as `lint` or `trailing_garbage`,
* apply any additional changes,
* update your `master` branch from upstream,
* rebase on `master` because there have been changes since 2019.

**Comment by DimitriPapadopoulos** [comment]
That being said, note that _json-parser_ is defined as a "**very low footprint JSON parser**", so I don't know if the changes required for the linter will be applied. I guess it depends on the footprint of the additional code. In any case the decision does not belong to me.

**Comment by tmegow** [comment]
@DimitriPapadopoulos I created #156 , may I request a review? Perhaps this draft can be closed.

**Comment by DimitriPapadopoulos** [comment]
Superseded by #156.

**Pull request closed without merging**